### ISSUE TITLE
PubSub: Calling iterator.Stop() goes into spinlock

### ISSUE THREAD

**Issue by aeijdenberg** [GoogleCloudPlatform/gcloud-golang#235]
The following code,which AFAICT is doing using the client library correctly, never completes:

	iter, _ := sub.Pull(ctx)
	message, _ := iter.Next()
	message.Done(true)
	
	fmt.Println("Pre done...")
	iter.Stop()
	fmt.Println("Post done (never reached)...")

Passing false (rather than true) to message.Done() makes the code complete.  Reading the docs, I'm not sure whether I'm meant to be passing true or false to indicate that I've completed processing, so perhaps that's a doc issue.

In any case, when passing true, the goroutine started in func (ka *keepAlive) Start() goes into a spinlock. Specifically this code:

    for {
        select {
            ...
            case <-ka.done: // since closed, always fires
                done = true
            ...
        }
        live, _ := ka.getAckIDs()
        if done && len(live) == 0 { // len(live) is not 0, so we loop again
            return
        }
    }

If running in GAE, this pretty much kills the rest of your process. :(

If we really are waiting on something (which by the way, I don't think we are, since it never completes even after I add sleep statements in), I suspect we should be blocking on something here...

**Comment by mcgreevy** [comment]
Yep, this is on my radar.  It should be fixed within the next day.

**Comment by mcgreevy** [comment]
@aeijdenberg I've sent out https://code-review.googlesource.com/4401 for review, which should fix this issue.  Feel free to patch it and let me know if it doesn't.

Passing false to Message.Done happens to avoid inducing this bug, but will not have the result you want: the message will be NACKed and subsequently redelivered.  I'll try to make this clearer in the docs.  In any case, once change 4401 is checked in, passing true will also function correctly.

**Comment by aeijdenberg** [comment]
Hi Michael, thanks for the patch. While it seems to stop the busy lock, calling `iter.Stop()` continues to hang for exactly 5 minutes on my system. A problem with the patch is that when `func (ka *keepAlive) Stop()` is called, `len(ka.items) == 0` is not true, and as such `ka.done` is not sent to and instead we block on `ka.wg.Wait()`.

Full test case below - perhaps I'm using the library wrong...

    package main

    import (
        "fmt"
        "golang.org/x/net/context"
        "google.golang.org/cloud/pubsub"
        "time"
    )

    const (
        PROJECT = "your-project-id-123456"
        TOPIC   = "top1"
        SUB     = "sub1"
    )

    func main() {
        ctx := context.Background()

        // Create client
        client, err := pubsub.NewClient(ctx, PROJECT)
        if err != nil {
            fmt.Println("Err1: %v", err)
            return
        }

        // Create topic if not exists
        topic := client.Topic(TOPIC)
        exists, err := topic.Exists(ctx)
        if err != nil {
            fmt.Println("Err7: %v", err)
            return
        }
        if !exists {
            topic, err = client.NewTopic(ctx, TOPIC)
            if err != nil {
                fmt.Println("Err2: %v", err)
                return
            }
        }

        // Create subscription if not exists
        sub := client.Subscription(SUB)
        exists, err = sub.Exists(ctx)
        if err != nil {
            fmt.Println("Err7: %v", err)
            return
        }
        if !exists {
            sub, err = topic.Subscribe(ctx, SUB, 10*time.Minute, nil)
            if err != nil {
                fmt.Println("Err3: %v", err)
                return
            }
        }

        // Publish to topic
        _, err = topic.Publish(ctx, &pubsub.Message{Data: []byte("foo")})
        if err != nil {
            fmt.Println("Err4: %v", err)
            return
        }

        // Get sub iterator
        iter, err := sub.Pull(ctx)
        if err != nil {
            fmt.Println("Err5: %v", err)
            return
        }

        // Pull one item
        message, err := iter.Next()
        if err != nil {
            fmt.Println("Err6: %v", err)
            return
        }

        // Mark it done
        message.Done(true)

        // Stop the iterator
        fmt.Println("Pre done...")
        iter.Stop()

        // Watch it burn
        fmt.Println("Post done...")
    }

**Comment by mcgreevy** [comment]
I suspect that the problem is that the batching of acknowledgements is introducing a delay here.  iter.Stop will block until the message has been acked on the server, which may take a while if the acknowledgement deadline for the subscription is long.  Do you have a long default ack deadline?  After I've submitted change 4401, I'm going to address this separate issue in a followup change.

**Comment by aeijdenberg** [comment]
Yes, I'd set 10 minutes here.

**Comment by mcgreevy** [comment]
I've opened issue #236 to track the related problem of Iterator.Stop not returning promptly.

**Comment by mcgreevy** [comment]
This should be all fixed now. @aeijdenberg  can you confirm that it's working for you?

**Comment by aeijdenberg** [comment]
I just tried to re-run my test-case and the latest version seems to have removed the TopicHandle.Subscribe() method (as confirmed by it's non-presence in the docs: https://godoc.org/google.golang.org/cloud/pubsub, even though the docs for CreateSub state "Deprecated: Use TopicHandle.Subscribe instead.Deprecated: Use TopicHandle.Subscribe instead.").  Old version deprecated, new one not ready yet? :)

**Comment by mcgreevy** [comment]
I recently replaced TopicHandle.Subscribe with Client.NewSubscription, so that it would be possible to create a subscription to a topic in another project.  I forgot to update the deprecation notice on CreateSub to point to the new method.  I'll update that. Thanks for pointing it out.

**Comment by aeijdenberg** [comment]
Thanks - I made the changes to my test case and it now terminates quickly as expected.
